File: examples/sdk_example.ts
/**
 * Sultan L1 SDK Example - TypeScript/JavaScript
 * 
 * This example demonstrates how to interact with Sultan L1 blockchain.
 * 
 * ## Installation
 * ```bash
 * npm install @noble/ed25519 bech32 fast-json-stable-stringify
 * ```
 * 
 * ## Usage
 * ```bash
 * npx ts-node sdk_example.ts
 * ```
 */

import * as ed25519 from '@noble/ed25519';
import { bech32 } from 'bech32';
import stringify from 'fast-json-stable-stringify';

const RPC_URL = 'https://rpc.sltn.io';

// ============================================================================
// HELPERS
// ============================================================================

/**
 * Convert bytes to hex string (required for signatures and public keys)
 */
function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
}

/**
 * Convert hex string to bytes
 */
function hexToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substr(i, 2), 16);
  }
  return bytes;
}

// ============================================================================
// WALLET
// ============================================================================

interface Wallet {
  privateKey: Uint8Array;
  publicKey: Uint8Array;
  address: string;
}

/**
 * Generate a new random wallet
 */
async function createWallet(): Promise<Wallet> {
  const privateKey = ed25519.utils.randomPrivateKey();
  const publicKey = await ed25519.getPublicKeyAsync(privateKey);
  
  // Derive address: SHA256(pubkey)[0:20] -> bech32("sultan")
  const hash = await crypto.subtle.digest('SHA-256', publicKey);
  const addressBytes = new Uint8Array(hash).slice(0, 20);
  const words = bech32.toWords(addressBytes);
  const address = bech32.encode('sultan', words);
  
  return { privateKey, publicKey, address };
}

/**
 * Import wallet from private key hex
 */
async function importWallet(privateKeyHex: string): Promise<Wallet> {
  const privateKey = hexToBytes(privateKeyHex);
  const publicKey = await ed25519.getPublicKeyAsync(privateKey);
  
  const hash = await crypto.subtle.digest('SHA-256', publicKey);
  const addressBytes = new Uint8Array(hash).slice(0, 20);
  const words = bech32.toWords(addressBytes);
  const address = bech32.encode('sultan', words);
  
  return { privateKey, publicKey, address };
}

// ============================================================================
// API TYPES
// ============================================================================

interface Balance {
  address: string;
  balance: number;
  nonce: number;
}

interface Status {
  node_id: string;
  block_height: number;
  validators: number;
  uptime_seconds: number;
  version: string;
  shard_count: number;
  tps_capacity: number;
}

interface TransactionResult {
  hash: string;
  from: string;
  to: string;
  amount: number;
  status: string;
  block_height?: number;
}

// ============================================================================
// SDK FUNCTIONS
// ============================================================================

/**
 * Get network status
 */
async function getStatus(): Promise<Status> {
  const res = await fetch(`${RPC_URL}/status`);
  return res.json();
}

/**
 * Get balance for an address
 */
async function getBalance(address: string): Promise<Balance> {
  const res = await fetch(`${RPC_URL}/balance/${address}`);
  return res.json();
}

/**
 * Get balance in human-readable SLTN
 */
async function getBalanceSltn(address: string): Promise<number> {
  const balance = await getBalance(address);
  return balance.balance / 1e9;
}

/**
 * Send SLTN tokens
 * 
 * CRITICAL: This function demonstrates proper signature format:
 * - Keys must be alphabetically sorted (use fast-json-stable-stringify)
 * - Amount must be a STRING in the signed message
 * - Signature and public key are hex encoded (NOT base64)
 */
async function sendSltn(
  wallet: Wallet,
  to: string,
  amountSltn: number
): Promise<TransactionResult> {
  // Get current nonce
  const balanceInfo = await getBalance(wallet.address);
  const nonce = balanceInfo.nonce;
  
  // Convert to atomic units
  const amountAtomic = Math.floor(amountSltn * 1e9);
  const timestamp = Math.floor(Date.now() / 1000);
  
  // Create message for signing
  // CRITICAL: Use fast-json-stable-stringify for alphabetical key order
  // CRITICAL: Amount must be a STRING
  const txForSigning = {
    amount: String(amountAtomic),  // MUST be string
    from: wallet.address,
    memo: '',
    nonce,
    timestamp,
    to
  };
  
  // Sign the message
  const message = new TextEncoder().encode(stringify(txForSigning));
  const signature = await ed25519.signAsync(message, wallet.privateKey);
  
  // Build request body
  const request = {
    tx: {
      from: wallet.address,
      to,
      amount: amountAtomic,  // Can be number in request body
      timestamp,
      nonce,
      memo: ''
    },
    signature: bytesToHex(signature),      // Hex encoded (128 chars)
    public_key: bytesToHex(wallet.publicKey) // Hex encoded (64 chars)
  };
  
  // Submit transaction
  const res = await fetch(`${RPC_URL}/tx`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(request)
  });
  
  return res.json();
}

/**
 * Get transaction by hash
 */
async function getTransaction(hash: string): Promise<TransactionResult> {
  const res = await fetch(`${RPC_URL}/tx/${hash}`);
  return res.json();
}

/**
 * Wait for transaction confirmation
 */
async function waitForConfirmation(hash: string, timeoutMs = 30000): Promise<TransactionResult> {
  const start = Date.now();
  
  while (Date.now() - start < timeoutMs) {
    const tx = await getTransaction(hash);
    if (tx.status === 'confirmed') {
      return tx;
    }
    // Wait 2 seconds (one block time)
    await new Promise(r => setTimeout(r, 2000));
  }
  
  throw new Error('Transaction confirmation timeout');
}

// ============================================================================
// EXAMPLE USAGE
// ============================================================================

async function main() {
  console.log('=== Sultan L1 SDK Example (TypeScript) ===\n');
  
  // Check network status
  console.log('📡 Checking network status...');
  const status = await getStatus();
  console.log(`   Block Height: ${status.block_height}`);
  console.log(`   Validators: ${status.validators}`);
  console.log(`   TPS Capacity: ${status.tps_capacity}`);
  console.log(`   Version: ${status.version}\n`);
  
  // Create a new wallet
  console.log('🔑 Creating new wallet...');
  const wallet = await createWallet();
  console.log(`   Address: ${wallet.address}`);
  console.log(`   Public Key: ${bytesToHex(wallet.publicKey).slice(0, 16)}...\n`);
  
  // Check balance
  console.log('💰 Checking balance...');
  const balance = await getBalanceSltn(wallet.address);
  console.log(`   Balance: ${balance} SLTN\n`);
  
  // Example: Send transaction (commented out - requires funded wallet)
  // console.log('📤 Sending 10 SLTN...');
  // const tx = await sendSltn(wallet, 'sultan1recipient...', 10);
  // console.log(`   TX Hash: ${tx.hash}`);
  // 
  // const confirmed = await waitForConfirmation(tx.hash);
  // console.log(`   Status: ${confirmed.status}`);
  
  console.log('✅ SDK example complete!');
  console.log('\nFor more examples, see: https://github.com/Sultan-Labs/DOCS/tree/main/examples');
}

// Run if executed directly
main().catch(console.error);

// ============================================================================
// EXPORTS (for use as a module)
// ============================================================================

export {
  createWallet,
  importWallet,
  getStatus,
  getBalance,
  getBalanceSltn,
  sendSltn,
  getTransaction,
  waitForConfirmation,
  bytesToHex,
  hexToBytes,
  Wallet,
  Balance,
  Status,
  TransactionResult
};

File: examples/sdk_example.py
#!/usr/bin/env python3
"""
Sultan L1 SDK Example - Python

This example demonstrates how to interact with Sultan L1 blockchain using Python.

Installation:
    pip install pynacl bech32 requests

Usage:
    python sdk_example.py
"""

import hashlib
import json
import time
from typing import Optional, Tuple

import requests
from bech32 import bech32_encode, convertbits
from nacl.signing import SigningKey
from nacl.encoding import RawEncoder

RPC_URL = "https://rpc.sltn.io"


# ============================================================================
# WALLET
# ============================================================================

class SultanWallet:
    """Sultan L1 Wallet for signing transactions."""
    
    def __init__(self, private_key_hex: Optional[str] = None):
        """
        Create a new wallet or import from private key.
        
        Args:
            private_key_hex: Optional hex-encoded private key. If None, generates new key.
        """
        if private_key_hex:
            self.signing_key = SigningKey(bytes.fromhex(private_key_hex))
        else:
            self.signing_key = SigningKey.generate()
        
        self.public_key = self.signing_key.verify_key.encode()
        self.address = self._derive_address()
    
    def _derive_address(self) -> str:
        """Derive bech32 address from public key."""
        # SHA-256 hash of public key, take first 20 bytes
        hash_bytes = hashlib.sha256(self.public_key).digest()[:20]
        # Bech32 encode with "sultan" prefix
        data = convertbits(list(hash_bytes), 8, 5)
        return bech32_encode("sultan", data)
    
    def sign(self, message: dict) -> Tuple[str, str]:
        """
        Sign a message dictionary.
        
        CRITICAL: Uses sorted keys for deterministic JSON ordering.
        
        Args:
            message: Dictionary to sign
            
        Returns:
            Tuple of (signature_hex, public_key_hex)
        """
        # CRITICAL: Use sorted keys for deterministic JSON
        msg_bytes = json.dumps(message, sort_keys=True).encode()
        signed = self.signing_key.sign(msg_bytes, encoder=RawEncoder)
        
        # IMPORTANT: Use hex encoding, NOT base64
        signature = signed.signature.hex()
        pubkey = self.public_key.hex()
        
        return signature, pubkey
    
    @property
    def private_key_hex(self) -> str:
        """Get private key as hex string (keep secret!)."""
        return self.signing_key.encode().hex()
    
    @property
    def public_key_hex(self) -> str:
        """Get public key as hex string."""
        return self.public_key.hex()


# ============================================================================
# SDK CLIENT
# ============================================================================

class SultanSDK:
    """Sultan L1 SDK Client."""
    
    def __init__(self, rpc_url: str = RPC_URL):
        """
        Initialize SDK client.
        
        Args:
            rpc_url: RPC endpoint URL
        """
        self.rpc_url = rpc_url
    
    @classmethod
    def mainnet(cls) -> "SultanSDK":
        """Create SDK instance for mainnet."""
        return cls("https://rpc.sltn.io")
    
    @classmethod
    def testnet(cls) -> "SultanSDK":
        """Create SDK instance for testnet."""
        return cls("https://testnet.sltn.io")
    
    def get_status(self) -> dict:
        """Get network status."""
        res = requests.get(f"{self.rpc_url}/status")
        res.raise_for_status()
        return res.json()
    
    def get_balance(self, address: str) -> dict:
        """
        Get balance for an address.
        
        Args:
            address: Sultan address (sultan1...)
            
        Returns:
            Dict with 'address', 'balance' (atomic units), 'nonce'
        """
        res = requests.get(f"{self.rpc_url}/balance/{address}")
        res.raise_for_status()
        return res.json()
    
    def get_balance_sltn(self, address: str) -> float:
        """
        Get balance in human-readable SLTN.
        
        Args:
            address: Sultan address (sultan1...)
            
        Returns:
            Balance in SLTN
        """
        data = self.get_balance(address)
        return data["balance"] / 1e9
    
    def send_sltn(
        self,
        wallet: SultanWallet,
        to: str,
        amount_sltn: float,
        memo: str = ""
    ) -> dict:
        """
        Send SLTN tokens.
        
        CRITICAL: This method demonstrates proper signature format:
        - Keys are alphabetically sorted (json.dumps with sort_keys=True)
        - Amount is a STRING in the signed message
        - Signature and public key are hex encoded (NOT base64)
        
        Args:
            wallet: Sender wallet
            to: Recipient address
            amount_sltn: Amount in SLTN
            memo: Optional memo
            
        Returns:
            Transaction result dict
        """
        # Get current nonce
        balance_info = self.get_balance(wallet.address)
        nonce = balance_info["nonce"]
        
        # Convert to atomic units
        amount_atomic = int(amount_sltn * 1e9)
        timestamp = int(time.time())
        
        # Create message for signing
        # CRITICAL: Amount must be a STRING
        tx_for_signing = {
            "amount": str(amount_atomic),  # MUST be string
            "from": wallet.address,
            "memo": memo,
            "nonce": nonce,
            "timestamp": timestamp,
            "to": to
        }
        
        # Sign with sorted keys
        signature, pubkey = wallet.sign(tx_for_signing)
        
        # Build request body
        request = {
            "tx": {
                "from": wallet.address,
                "to": to,
                "amount": amount_atomic,  # Can be int in request body
                "timestamp": timestamp,
                "nonce": nonce,
                "memo": memo
            },
            "signature": signature,    # Hex encoded (128 chars)
            "public_key": pubkey       # Hex encoded (64 chars)
        }
        
        # Submit transaction
        res = requests.post(
            f"{self.rpc_url}/tx",
            json=request,
            headers={"Content-Type": "application/json"}
        )
        res.raise_for_status()
        return res.json()
    
    def get_transaction(self, tx_hash: str) -> dict:
        """
        Get transaction by hash.
        
        Args:
            tx_hash: Transaction hash
            
        Returns:
            Transaction details
        """
        res = requests.get(f"{self.rpc_url}/tx/{tx_hash}")
        res.raise_for_status()
        return res.json()
    
    def wait_for_confirmation(
        self,
        tx_hash: str,
        timeout_seconds: int = 30
    ) -> dict:
        """
        Wait for transaction confirmation.
        
        Args:
            tx_hash: Transaction hash
            timeout_seconds: Maximum wait time
            
        Returns:
            Confirmed transaction
            
        Raises:
            TimeoutError: If transaction not confirmed in time
        """
        start = time.time()
        
        while time.time() - start < timeout_seconds:
            tx = self.get_transaction(tx_hash)
            if tx.get("status") == "confirmed":
                return tx
            time.sleep(2)  # Wait one block time
        
        raise TimeoutError("Transaction confirmation timeout")


# ============================================================================
# EXAMPLE USAGE
# ============================================================================

def main():
    print("=== Sultan L1 SDK Example (Python) ===\n")
    
    # Initialize SDK
    sdk = SultanSDK.mainnet()
    
    # Check network status
    print("📡 Checking network status...")
    status = sdk.get_status()
    print(f"   Block Height: {status['block_height']}")
    print(f"   Validators: {status['validators']}")
    print(f"   TPS Capacity: {status['tps_capacity']}")
    print(f"   Version: {status['version']}\n")
    
    # Create a new wallet
    print("🔑 Creating new wallet...")
    wallet = SultanWallet()
    print(f"   Address: {wallet.address}")
    print(f"   Public Key: {wallet.public_key_hex[:16]}...\n")
    
    # Check balance
    print("💰 Checking balance...")
    balance = sdk.get_balance_sltn(wallet.address)
    print(f"   Balance: {balance} SLTN\n")
    
    # Example: Send transaction (commented out - requires funded wallet)
    # print("📤 Sending 10 SLTN...")
    # tx = sdk.send_sltn(wallet, "sultan1recipient...", 10.0)
    # print(f"   TX Hash: {tx['hash']}")
    #
    # confirmed = sdk.wait_for_confirmation(tx['hash'])
    # print(f"   Status: {confirmed['status']}")
    
    print("✅ SDK example complete!")
    print("\nFor more examples, see: https://github.com/Sultan-Labs/DOCS/tree/main/examples")


if __name__ == "__main__":
    main()
